cluster-autoscaler/cloudprovider/kubemark/kubemark_linux.go

```diff
@@ -14,16 +16,243 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
-// +build linux
-
 // Version to be compiled in the linux environment. May cause compilation issues on
 // other OS.
 
 package kubemark
 
 import (
-	// Placeholder
-	_ "k8s.io/kubernetes/pkg/kubemark"
-	// Placeholder
-	_ "k8s.io/kubernetes/plugin/pkg/scheduler/schedulercache"
+	"fmt"
+
+	apiv1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/autoscaler/cluster-autoscaler/cloudprovider"
+	"k8s.io/autoscaler/cluster-autoscaler/config/dynamic"
+	"k8s.io/autoscaler/cluster-autoscaler/utils/errors"
+	"k8s.io/kubernetes/pkg/kubemark"
+	"k8s.io/kubernetes/plugin/pkg/scheduler/schedulercache"
+
+	"github.com/golang/glog"
+)
+
+const (
+	// ProviderName is the cloud provider name for kubemark
+	ProviderName = "kubemark"
 )
+
+// KubemarkCloudProvider implements CloudProvider interface for kubemark
+type KubemarkCloudProvider struct {
+	kubemarkController *kubemark.KubemarkController
+	nodeGroups         []*NodeGroup
+}
+
+// BuildKubemarkCloudProvider builds a CloudProvider for kubemark. Builds
+// node groups from passed in specs.
+func BuildKubemarkCloudProvider(kubemarkController *kubemark.KubemarkController, specs []string) (*KubemarkCloudProvider, error) {
+	kubemark := &KubemarkCloudProvider{
+		kubemarkController: kubemarkController,
+		nodeGroups:         make([]*NodeGroup, 0),
+	}
+	for _, spec := range specs {
+		if err := kubemark.addNodeGroup(spec); err != nil {
+			return nil, err
+		}
+	}
+	return kubemark, nil
+}
+
+func (kubemark *KubemarkCloudProvider) addNodeGroup(spec string) error {
+	nodeGroup, err := buildNodeGroup(spec, kubemark.kubemarkController)
+	if err != nil {
+		return err
+	}
+	glog.V(2).Infof("adding node group: %s", nodeGroup.Name)
+	kubemark.nodeGroups = append(kubemark.nodeGroups, nodeGroup)
+	return nil
+}
+
+// Name returns name of the cloud provider.
+func (kubemark *KubemarkCloudProvider) Name() string {
+	return ProviderName
+}
+
+// NodeGroups returns all node groups configured for this cloud provider.
+func (kubemark *KubemarkCloudProvider) NodeGroups() []cloudprovider.NodeGroup {
+	result := make([]cloudprovider.NodeGroup, 0, len(kubemark.nodeGroups))
+	for _, nodegroup := range kubemark.nodeGroups {
+		result = append(result, nodegroup)
+	}
```

Comment:
golang trivia - you can avoid the loop by doing `result = append(result, kubemark.nodeGroups...)`